### PR TITLE
Use unit test variant's runtimeConfiguration instead of main variant's

### LINKED ISSUE #1968: `ClassNotFoundException` when upgrading `androidx.navigation` from `2.8.9` to `2.9.0`
**Description**

When upgrading `androidx.navigation` from `2.8.9` to `2.9.0` I get this error. 

It looks like `androidx.navigation` removed a transitive dep on `androidx.navigation:navigation-common-ktx` in the new version. 

 `navigation-common-ktx-2.9.0.aar` is essentially empty with no classes (no R class) 

Not sure why paparazzi is still trying to load this class that no longer exists. 

```
    java.lang.ClassNotFoundException: androidx.navigation.common.ktx.R
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:421)
        at java.base/java.lang.Class.forName(Class.java:412)
        at app.cash.paparazzi.internal.PaparazziCallback.initResources(PaparazziCallback.kt:59)
        at app.cash.paparazzi.PaparazziSdk.prepare(PaparazziSdk.kt:124)
        at app.cash.paparazzi.Paparazzi.prepare(Paparazzi.kt:89)
        at app.cash.paparazzi.Paparazzi$apply$1.evaluate(Paparazzi.kt:77)
        at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
        at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
        at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
```

**Expected behavior**

Don't try to load `R` classes that don't exist. 

**Additional information:**
- Paparazzi Version: 1.3.5

### LINKED ISSUE #2286: ClassNotFoundException for AndroidX "anchor" module R classes in initResources
## Description

After updating AndroidX dependencies (DataStore 1.1.x+, Lifecycle 2.10.x+), Paparazzi tests crash with `ClassNotFoundException` because `PaparazziCallback.initResources()` tries to load R classes for AndroidX **anchor modules** that don't have any.

Anchor modules (e.g. `datastore-preferences-android`, `lifecycle-viewmodel-android`) are empty marker artifacts used by AndroidX for dependency resolution. They declare a package name (e.g. `androidx.datastore.datastore.preferences.anchor`) which ends up in `resources.json` via `PrepareResourcesTask`, but they don't generate an R class.

## Stack trace

`java.lang.ClassNotFoundException: androidx.datastore.datastore.preferences.anchor.R at java.base/java.lang.Class.forNameClass.java:377 at app.cash.paparazzi.internal.PaparazziCallback.initResourcesPaparazziCallback.kt:59 at app.cash.paparazzi.PaparazziSdk.preparePaparazziSdk.kt:122 at app.cash.paparazzi.Paparazzi.preparePaparazzi.kt:86 at app.cash.paparazzi.Paparazziapply1.evaluatePaparazzi.kt:74`

## Root cause

In [`PaparazziCallback.initResources()`](https://github.com/cashapp/paparazzi/blob/master/paparazzi/src/main/java/app/cash/paparazzi/internal/PaparazziCallback.kt), `Class.forName("$rPackageName.R")` is called without a try-catch for `ClassNotFoundException`:

```kotlin
fun initResources() {
    for (rPackageName in resourcePackageNames) {
        val rClass = Class.forName("$rPackageName.R") // crashes if R class doesn't exist
        // ...
    }
}
```

The problematic package names in `resources.json` (generated by `PrepareResourcesTask`):
- `androidx.datastore.datastore.preferences.anchor`
- `androidx.lifecycle.lifecycle.viewmodel.anchor`

## Expected behavior

Paparazzi should gracefully skip packages that don't have an R class, instead of crashing.

## Suggested fix

Either:

1. **In `PaparazziCallback.initResources()`** — wrap `Class.forName` in a try-catch and `continue`:
```kotlin
for (rPackageName in resourcePackageNames) {
    val rClass = try {
        Class.forName("$rPackageName.R")
    } catch (_: ClassNotFoundException) {
        continue
    }
    // ...
}
```

2. **In `PrepareResourcesTask`** — filter out package names ending with `.anchor` from `resourcePackageNames` before writing `resources.json`.

## Workaround

We work around this by adding a `doLast` action to the `preparePaparazzi*Resources` task that filters anchor packages from the generated `resources.json`:

```kotlin
tasks.matching { it.name.startsWith("preparePaparazzi") && it.name.endsWith("Resources") }
    .configureEach {
        doLast {
            val resourcesFile = outputs.files.singleFile
            if (!resourcesFile.exists()) return@doLast
            // filter out entries ending with ".anchor" from resourcePackageNames in the JSON
        }
    }
```

## Environment

- **Paparazzi**: 2.0.0-alpha02 (also reproduced by inspecting alpha04 bytecode — same code path)
- **AGP**: 8.13.2
- **Gradle**: 9.2.1
- **AndroidX DataStore**: 1.1.x (transitive via Firebase BOM 34.9.0)
- **AndroidX Lifecycle**: 2.10.0

### PR DESCRIPTION
**Pull request by SimonMarquis**
This ensures only the resolved runtime **unit test** dependencies are used.

image

Fixes #1968
Fixes #2286

### CODE REVIEW

File: paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt [hunk 1]
```diff
@@ -945,6 +945,7 @@ class PaparazziPluginTest {
     val config = resourcesFile.loadConfig()
     assertThat(config.mainPackage).isEqualTo("app.cash.paparazzi.plugin.test")
     assertThat(config.resourcePackageNames).containsExactly(
+      "app.cash.paparazzi.plugin.test",
       "app.cash.paparazzi.plugin.test",
```
Comment (SimonMarquis) [inline paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt]: Should we use a `buildSet {...}` to avoid duplicated values that will result in the same corresponding `R` class?

https://github.com/cashapp/paparazzi/blob/c640cbbdddae046dcc7868c729867630f491834e/paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PrepareResourcesTask.kt#L73-L78

File: paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt [hunk 2]
```diff
@@ -956,6 +957,7 @@ class PaparazziPluginTest {
       "build/generated/res/extra"
     )
     assertThat(config.moduleResourceDirs).containsExactly(
+      "build/intermediates/packaged_res/debug/packageDebugResources",
```
Comment (SimonMarquis) [inline paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt:959]: Any comments about these additions in the `moduleResourceDirs` & `projectAssetDirs` that are now also linking the consumer project dirs?

Comment (SimonMarquis) [reply paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt:959]: Is this because the 'main' variant is now seen as a 'dependency'?

File: paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt (hunk 1 above)
Comment (colinmarsch) [reply paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt]: Deduping seems reasonable here to me. Maybe we can use `distinct` or `distinctBy` on the `buildList` to keep the type the same for less change here?

**Review by geoff-powell — APPROVED** [review]

File: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/AndroidVariantSources.kt [hunk 3]
```diff
@@ -1,8 +1,10 @@
 package app.cash.paparazzi.gradle
 
 import app.cash.paparazzi.gradle.utils.artifactsFor
+import com.android.build.api.variant.UnitTest
 import com.android.build.api.variant.Variant
 import com.android.build.gradle.internal.publishing.AndroidArtifacts
+import org.gradle.api.artifacts.Configuration
```
Comment (colinmarsch) [inline paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/AndroidVariantSources.kt]: nit: this is unused

File: paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt (hunk 2 above)
Comment (colinmarsch) [reply paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt]: > Is this because the 'main' variant is now seen as a 'dependency'?

Yeah, the unit test runtime configuration includes the main component. Could we filter out the current component in this change to avoid this duplication?

Comment (SimonMarquis) [reply paparazzi-gradle-plugin/src/test/java/app/cash/paparazzi/gradle/PaparazziPluginTest.kt]: Done in https://github.com/cashapp/paparazzi/pull/2362/commits/246127648b6d7fc566ae06dd44fc2298858bba4b

**Comment by SimonMarquis** [comment]
@colinmarsch the review comments have been addressed, PTAL 🙏

File: paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/AndroidVariantSources.kt [hunk 4]
```diff
@@ -12,21 +13,25 @@ import org.gradle.api.provider.Provider
...
     variant.sources.res?.all?.map { layers -> layers.flatten() }?.map { it.asReversed() }
   }
 
   // https://android.googlesource.com/platform/tools/base/+/96015063acd3455a76cdf1cc71b23b0828c0907f/build-system/gradle-core/src/main/java/com/android/build/gradle/tasks/MergeResources.kt#875
   val moduleResourceDirs: FileCollection by lazy {
-    variant.runtimeConfiguration
-      .artifactsFor(AndroidArtifacts.ArtifactType.ANDROID_RES.type) { it is ProjectComponentIdentifier }
+    unitTest.runtimeConfiguration
+      .artifactsFor(AndroidArtifacts.ArtifactType.ANDROID_RES.type) {
+        it is ProjectComponentIdentifier && it.projectPath != projectPath
```
Comment (colinmarsch) [inline paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/AndroidVariantSources.kt]: Could we distinguish the containing build as well as the `projectPath` here? `projectPath` is only unique within a build, so this could incorrectly exclude an included-build dependency whose path matches the consumer project, for example if both are `:library`. `buildTreePath` would address this on Gradle 8.3+, though we may need a compatibility-safe equivalent for older supported Gradle versions. Same with the two places below.

Comment (SimonMarquis) [reply paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/AndroidVariantSources.kt]: Comparing `Project.path` and `Gradle.buildPath` seems to be good enough.
Let me know if using a `Project` reference like this is banned, but as far as i check, values are not ending up in the configuration cache 👍

Comment (SimonMarquis) [reply paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/AndroidVariantSources.kt]: Actually, forcing these properties in the unit tests, brings up CC errors.
```
org.gradle.configuration-cache=true
org.gradle.configuration-cache.problems=fail
```



I fixed them in https://github.com/cashapp/paparazzi/pull/2362/commits/f46734d4b00299050a1e85a7f50de8965cf64d58, but at the same time I'm wondering why we have all these `by lazy`? Removing them would allow us to not expose `Variant`, `UnitTest`, `Project` as vals of `AndroidVariantSources`, making it CC-leak safe.
Also, I feel like `ArtifactCollection` -> `FileCollection` are somewhat lazy APIs, but I might be wrong.

**Comment by colinmarsch** [comment]
Could we add a regression test here for the #2286 case? I think the empty anchor AAR issue isn't actually resolved by this change.

**Comment by SimonMarquis** [comment]
@colinmarsch I can, but I'm not able to reproduce the issue on master with this test `build.gradle` file inside `paparazzi-gradle-plugin`:

```
plugins {
  id 'com.android.library'
  id 'app.cash.paparazzi'
}

android {
  namespace = 'app.cash.paparazzi.plugin.test'
  compileSdk = libs.versions.compileSdk.get() as int
  defaultConfig {
    minSdk = libs.versions.minSdk.get() as int
  }
  compileOptions {
    sourceCompatibility = libs.versions.javaTarget.get()
    targetCompatibility = libs.versions.javaTarget.get()
  }
}

dependencies {
  implementation 'androidx.lifecycle:lifecycle-viewmodel-android:2.10.0'
  implementation 'androidx.datastore:datastore-preferences-android:1.1.7'
  implementation 'com.squareup.okhttp3:okhttp:5.1.0'
}
```

I tested the 3 mentioned libraries (although there was no complete artifact coordinate provided) and the tests just works.
